gh-111520: Integrate the Tier 2 interpreter in the Tier 1 interpreter

gvanrossum opened the pull request:
See also https://github.com/faster-cpython/ideas/issues/631

**Beware!** This changes the env vars to enable uops and their debugging to `PYTHON_UOPS` and `PYTHON_LLTRACE`.

<!-- gh-issue-number: gh-111520 -->
* Issue: gh-111520
<!-- /gh-issue-number -->

gvanrossum committed:
DO NOT MERGE: Always use -Xuops

gvanrossum committed:
Make all labels in _PyUopExecute end in _tier_two

Use `GOTO_ERROR(error)` instead of `goto error`.
This macro is defined differently in the tier two interpreter.

gvanrossum committed:
Control lltrace via PYTHON_LLTRACE=N

This replaces PYTHONUOPSDEBUG=N. The meaning of N is the same (for now):

  0: no tracing
  1: print when tier 2 trace created
  2: print contents of tier 2 trace
  3: print every uop executed
  4: print optimization attempts and details
  5: print tier 1 instructions and stack

gvanrossum committed:
Rename PYTHONUOPS to PYTHON_UOPS for consistency

gvanrossum committed:
Integrate Tier 2 into _PyEval_EvalFrameDefault

markshannon commented on Python/bytecodes.c:
```diff
@@ -2318,13 +2318,16 @@ dummy_func(
             Py_INCREF(executor);
+            if (executor->execute == _PyUopExecute) {
+                self = (_PyUOpExecutorObject *)executor;
+                goto enter_tier_two;
```
`bytecodes.c` should describe the semantics, so maybe add a `ENTER_TIER_TWO` macro, which can be replaced by the code generator?

gvanrossum replied:
Okay, naming it `GOTO_TIER_TWO()`.

markshannon commented on Python/ceval.c:
```diff
@@ -942,6 +945,123 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+    #define DPRINTF(level, ...)
+#endif
+
+    CHECK_EVAL_BREAKER();
```
This shouldn't be here. All eval breaker check are explicit in tier 2.

gvanrossum replied:
Okay, deleting.

markshannon commented on Python/ceval.c:
```diff
@@ -942,6 +945,123 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+    CHECK_EVAL_BREAKER();
+
+    OPT_STAT_INC(traces_executed);
+    _Py_CODEUNIT *ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
```
If we use an instruction pointer, we shouldn't need this?

gvanrossum replied:
I will inline it.

markshannon commented on Python/ceval.c:
```diff
@@ -942,6 +945,123 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+
+    OPT_STAT_INC(traces_executed);
+    _Py_CODEUNIT *ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
+    _PyUOpInstruction *next_uop = self->trace;
```
This should have been set by `ENTER_EXECUTOR`.

gvanrossum replied:
Yes and no. It would mean that `next_uop` would have to be a local in the top-level function scope. I am trying to reduce the number of locals there. For now let me keep it this way. I expect that it will be fine even when we start stitching, because each trace can only be entered at the top and we need the executor (so that we can decref it upon exiting the trace). The initial `next_uop` is just the executor plus a fixed offset anyway.

markshannon commented on Python/ceval.c:
```diff
@@ -942,6 +945,123 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+#define TIER_TWO 2
+#include "executor_cases.c.h"
+
+            default:
```
Make this `Py_UNREACHABLE` for non-debug builds.

gvanrossum replied:
Done.

markshannon commented on Python/ceval.c:
```diff
@@ -942,6 +945,123 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+    );
+    goto error_tier_two;
+
+pop_4_error_tier_two:
```
Can't this be shared with tier 1? Unwinding should work the same.

gvanrossum replied:
Alas, not quite. The debug output is different, the stats collection is different, but most importantly, we need to `DECREF(self)` here before jumping to error.

gvanrossum replied:
Also we need to set `next_instr = frame->instr_ptr`.

markshannon reviewed (commented):
Looks promising.

gvanrossum committed:
Merge branch 'main' into mix-tiers

gvanrossum committed:
Merge branch 'main' into mix-tiers

gvanrossum committed:
Get rid of separate executor.c file

gvanrossum commented on Python/ceval.c:
```diff
@@ -942,6 +945,123 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+    frame->return_offset = 0;  // Don't leave this random
+    _PyFrame_SetStackPointer(frame, stack_pointer);
+    Py_DECREF(self);
+    goto resume_with_error;
```
We may not need to reset `return_offset`; we can also skip syncing `stack_pointer`; so we could make this slightly faster as follows (but it doesn't matter since errors are presumed to be on the slow path):
```suggestion
    Py_DECREF(self);
    next_instr = frame->instr_ptr;
    goto error;
```

gvanrossum committed:
Most suggestions from Mark's code review

- Remove CHECK_EVAL_BREAKER() from top of Tier 2 loop
- Make Tier 2 default case Py_UNREACHABLE() in non-debug mode
- GOTO_TIER_TWO() macro instead of `goto enter_tier_two`
- Inline ip_offset (Tier 2 LOAD_IP() is now a no-op)
- Move #define/#under TIER_ONE/TIER_TWO into generated .c.h files
- ceval_macros.h defines Tier 1 macros, Tier 2 is inlined in ceval.c

gvanrossum committed:
Fix test_generated_cases.py by stripping preprocessor prefix/suffix

gvanrossum committed:
Eradicate executors.c from Windows build files

gvanrossum reviewed (commented):
I consider this sufficiently cleaned up to undraft it, and if @markshannon doesn't comment I'll merge tomorrow.

I could noodle endlessly with the code for dropping from tier 2 back into tier 1, with and without errors; but it may be better to put that off until a more careful review.

gvanrossum committed:
Rename deoptimize_tier_two back to deoptimize (for Justin)

gvanrossum committed:
Fix whitespace

gvanrossum committed:
Revert "DO NOT MERGE: Always use -Xuops"

(I accidentally kept this commit after pushing it experimentally.
This means I've been testing with uops on all the time, which is
actually pretty amazing.)

This reverts commit e0e60ced06794dba547a03bf61729484eb361f53.

gvanrossum committed:
Add blurb

gvanrossum committed:
Add more color to the news blurb

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into mix-tiers

gvanrossum committed:
Eliminate 'operand' local variable

gvanrossum committed:
Rename self -> current_executor (TODO: eliminate it?)

gvanrossum committed:
Move `_EXIT_TRACE` logic to a separate label

Using `GOTO_TIER_ONE()` macro. This should make things simpler
for Justin.

gvanrossum committed:
Limit infinite recursion in test_typing

gvanrossum committed:
Limit infinite recursion in test_fileio

gvanrossum committed:
Limit infinite recursion in test_xml_etree

gvanrossum committed:
Limit infinite recursion in test_call

gvanrossum committed:
Fix test_call better: adjust Py_C_RECURSION_LIMIT in pystate.h

gvanrossum committed:
Revert unnecessary fixes to recursive tests

gvanrossum committed:
Even better fix -- increase stack space on Windows in debug mode
